src/xdp/program.c

```diff
@@ -692,7 +692,7 @@ XdpParseFrame(
         }
         Cache->Ip4Hdr = (IPV4_HEADER *)&Va[Offset];
         Cache->Ip4Valid = TRUE;
-        Offset += sizeof(*Cache->Ip4Hdr);
+        Offset += (((UINT32) Cache->Ip4Hdr->HeaderLength) << 2);
```

Comment:
This needs to make sure `Buffer->DataLength` actually has space.

Reply:
```c
        if (Buffer->DataLength < Offset + (((UINT32)Cache->Ip4Hdr->HeaderLength) << 2)) {
            goto BufferTooSmall;
        }
```
This needs to be added after the other size check above.

Reply:
Just re-pushed with this fix

Reply:
You can't simply replace the check, because you're reading from the memory by reading `Ip4Hdr->HeaderLength`. You have to first check the length is long enough to read that, and then you can read it.

Reply:
@mtfriesen do we have a test that fuzzes the receive path with various length packets and detects out-of-bounds reads?

Reply:
We should open a bug to (eventually) close this gap.

Reply:
I opened https://github.com/microsoft/xdp-for-windows/issues/185.